src/adapter/src/frontend_peek.rs

```diff
@@ -504,14 +560,52 @@ impl PeekClient {
                         real_time_recency_ts,
                     )
                     .await?;
+
+                // If this is the first (non-AS OF) query in a multi-statement transaction, store
+                // the read holds in the coordinator, so subsequent queries can validate against
+                // them.
+                if in_immediate_multi_stmt_txn {
+                    self.call_coordinator(|tx| Command::StoreTransactionReadHolds {
+                        conn_id: session.conn_id().clone(),
+                        read_holds: read_holds.clone(),
+                        tx,
+                    })
+                    .await;
+                }
+
                 (determination, read_holds)
             }
         };
 
-        // The old peek sequencing's sequence_peek_timestamp does two more things here:
-        // the txn_read_holds stuff and session.add_transaction_ops. We do these later in the new
-        // peek sequencing code, because at this point we might still bail out from the new peek
-        // sequencing, in which case we don't want the mentioned side effects to happen.
```

Comment:
We don't have any bailouts left after this point anymore, so it's now ok to do these things here.